core/cap-0063.md

```diff
@@ -0,0 +1,496 @@
++    int64 feeReadLiveSorobanEntry;  // Fee per live soroban ledger entry read
++    int64 feeWriteLedgerEntry;      // Fee per ledger entry write
+ 
+-    int64 feeRead1KB;  // Fee for reading 1KB
++    int64 feeReadLiveSoroban1KB;    // Fee for reading 1KB of live Soroban state
+ 
+     // The following parameters determine the write fee per 1KB.
+     // Write fee grows linearly until bucket list reaches this size
+@@ -60,6 +60,26 @@ struct ConfigSettingContractLedgerCostV0
+     uint32 bucketListWriteFeeGrowthFactor;
+ };
+ 
++struct ConfigSettingContractLedgerCostExtV1
++{
++    // Maximum number of evicted Soroban and classic entry read operations per ledger
++    uint32 ledgerMaxReadEvictedAndClassicEntries;
++
++    // Maximum number of bytes of evicted Soroban and classic state that can be
++    // read per ledger
++    uint32 ledgerMaxReadEvictedAndClassicBytes;
++
++    // Maximum number of evicted Soroban and classic entry read operations per transaction
++    uint32 txMaxReadEvictedAndClassicEntries;
++
++    // Maximum number of bytes of evicted Soroban and classic state that can be
++    // read per transaction
++    uint32 txMaxReadEvictedAndClassicBytes;
++
++    int64 feeReadEvictedAndClassicEntry;  // Fee per evicted Soroban/classic entry read
++    int64 feeReadEvictedAndClassic1KB;    // Fee for reading 1 KB of evicted Soroban/classic state
++};
++
+ // Historical data (pushed to core archives) settings for contracts.
+ struct ConfigSettingContractHistoricalDataV0
+ {
+@@ -302,7 +322,8 @@ enum ConfigSettingID
+     CONFIG_SETTING_STATE_ARCHIVAL = 10,
+     CONFIG_SETTING_CONTRACT_EXECUTION_LANES = 11,
+     CONFIG_SETTING_BUCKETLIST_SIZE_WINDOW = 12,
+-    CONFIG_SETTING_EVICTION_ITERATOR = 13
++    CONFIG_SETTING_EVICTION_ITERATOR = 13,
++    CONFIG_SETTING_CONTRACT_LEDGER_COST_EXT_V1 = 14
+ };
+ 
+ union ConfigSettingEntry switch (ConfigSettingID configSettingID)
+@@ -335,5 +356,7 @@ case CONFIG_SETTING_BUCKETLIST_SIZE_WINDOW:
+     uint64 bucketListSizeWindow<>;
+ case CONFIG_SETTING_EVICTION_ITERATOR:
+     EvictionIterator evictionIterator;
++case CONFIG_SETTING_CONTRACT_LEDGER_COST_EXT_V1:
++    ConfigSettingContractLedgerCostExtV1 ledgerCostExtV1;
+ };
+ }
+diff --git a/Stellar-transaction.x b/Stellar-transaction.x
+index 7d32481..0abe3e4 100644
+--- a/Stellar-transaction.x
++++ b/Stellar-transaction.x
+@@ -888,10 +888,31 @@ struct SorobanResources
+     uint32 writeBytes;
+ };
+ 
++struct SorobanResourcesExtV1
++{
++    // The maximum number of bytes of evicted Soroban state
++    // and classic state that can be read
++    uint32 readEvictedAndClassicBytes;
++
++    // Bitmap representing what Soroban entries in the
++    // footprint are evicted, based on the order of keys
++    // provided in the LedgerFootprint vectors. readOnly
++    // vector is first, followed by readWrite.
++    // If no keys are evicted, BitMap can be empty.
++    BitMap evictedSorobanEntries;
```

Comment:
We don't need hacky indexing here because only RW entries can be restored.

Also, I've mentioned this already offline, I think we should have a vec of indices of entries that need to be restored.  This will reduce the size of 99% of transactions as almost no transactions need to restore any entries. I think a slight increase in size for the txs that do restore entries is worth it. Also, it will be easier to interpret a sequence of indicies vs an opaque blob.

Reply:
I agree with @dmkozh I think a separate list of just the evicted entries is easier to understand and will be more space efficient in the average case

Reply:
Did a little digging and we're currently averaging 1.367 keys per restore op. I thought it would be way higher so I agree, vector of indices is better. Should we go uint8 or uint16? Do we ever forsee footprints going beyond 255?

Reply:
I'd go for u16 just in case. The overhead is marginal, but I'd say accessing 255+ keys is not out of the realm of possibility Though restoring 255+ entries seems unlikely... Still, I feel like it might create unnecessary confusion in the future.

Reply:
Are you planning to have 16-bit integer values in the XDR? XDR doesn't support a 16-bit integer type, the smallest integer is a 32-bit, because all values in XDR are padded to 4 bytes, so a 1 or 2 byte value would be 4 bytes anyway.